File: javascript/libjoynr-js/src/test/js/joynr/provider/TestProvider.js
/*global joynrTestRequire: true */

/*
 * #%L
 * %%
 * Copyright (C) 2011 - 2015 BMW Car IT GmbH
 * %%
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * 
 *      http://www.apache.org/licenses/LICENSE-2.0
 * 
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 * #L%
 */

joynrTestRequire("joynr/provider/TestProvider", [
    "joynr/vehicle/RadioProvider",
    "joynr/datatypes/exampleTypes/Country",
    "joynr/provider/ProviderAttributeNotifyReadWrite",
    "joynr/provider/ProviderAttributeNotifyRead",
    "joynr/provider/ProviderAttributeNotifyWrite",
    "joynr/provider/ProviderAttributeNotify",
    "joynr/provider/ProviderAttributeReadWrite",
    "joynr/provider/ProviderAttributeRead",
    "joynr/provider/ProviderAttributeWrite",
    "joynr/provider/ProviderOperation",
    "joynr/provider/ProviderEvent",
    "joynr/TypesEnum",
    "joynr/util/uuid"
], function(
        RadioProvider,
        Country,
        ProviderAttributeNotifyReadWrite,
        ProviderAttributeNotifyRead,
        ProviderAttributeNotifyWrite,
        ProviderAttributeNotify,
        ProviderAttributeReadWrite,
        ProviderAttributeRead,
        ProviderAttributeWrite,
        ProviderOperation,
        ProviderEvent,
        TypesEnum,
        uuid) {
    describe("libjoynr-js.joynr.provider.Provider", function() {
        var implementation = null;
        var dependencies = {
            ProviderAttributeNotifyReadWrite : ProviderAttributeNotifyReadWrite,
            ProviderAttributeNotifyRead : ProviderAttributeNotifyRead,
            ProviderAttributeNotifyWrite : ProviderAttributeNotifyWrite,
            ProviderAttributeNotify : ProviderAttributeNotify,
            ProviderAttributeReadWrite : ProviderAttributeReadWrite,
            ProviderAttributeRead : ProviderAttributeRead,
            ProviderAttributeWrite : ProviderAttributeWrite,
            ProviderOperation : ProviderOperation,
            ProviderEvent : ProviderEvent,
            TypesEnum : TypesEnum,
            uuid : uuid
        };

        beforeEach(function() {
            implementation =
                    {
                        isOn : {
                            value : false,
                            get : function() {
                                return this.value;
                            },
                            set : function(newValue) {
                                this.value = newValue;
                            }
                        },
                        numberOfStations : {
                            value : 0,
                            get : function() {
                                return this.value;
                            },
                            set : function(newValue) {
                                this.value = newValue;
                            }
                        },
                        mixedSubscriptions : {
                            value : "testvalue",
                            get : function() {
                                return this.value;
                            },
                            set : function(newValue) {
                                this.value = newValue;
                            }
                        },
                        enumAttribute : {
                            value : Country.GERMANY,
                            get : function() {
                                return this.value;
                            },
                            set : function(newValue) {
                                this.value = newValue;
                            }
                        },
                        enumArrayAttribute : {
                            value : [ Country.GERMANY
                            ],
                            get : function() {
                                return this.value;
                            },
                            set : function(newValue) {
                                this.value = newValue;
                            }
                        },
                        attrProvidedImpl : {
                            value : "testValue2",
                            get : function() {
                                return this.value;
                            },
                            set : function(newValue) {
                                this.value = newValue;
                            }
                        },
                        addFavoriteStation : jasmine.createSpy("addFavoriteStation"),
                        weakSignal : jasmine.createSpy("weakSignal"),
                        methodProvidedImpl : jasmine.createSpy("methodProvidedImpl"),
                        operationWithEnumsAsInputAndOutput : jasmine
                                .createSpy("operationWithEnumsAsInputAndOutput")
                    };
        });

        it(
                "RadioProvider does not throw when instantiated with different implementations",
                function() {
                    expect(function() {
                        var radioProvider = new RadioProvider({}, dependencies);
                    }).not.toThrow();

                    expect(function() {
                        var radioProvider = new RadioProvider({
                            isOn : {}
                        }, dependencies);
                    }).not.toThrow();

                    expect(function() {
                        var radioProvider = new RadioProvider({
                            isOn : {
                                get : function() {
                                    return false;
                                }
                            }
                        }, dependencies);
                    }).not.toThrow();

                    expect(function() {
                        var radioProvider = new RadioProvider({
                            isOn : {
                                set : function(value) {}
                            }
                        }, dependencies);
                    }).not.toThrow();

                    expect(function() {
                        var radioProvider = new RadioProvider({
                            isOn : {
                                get : function() {
                                    return false;
                                },
                                set : function(value) {}
                            }
                        }, dependencies);
                    }).not.toThrow();

                    expect(function() {
                        var radioProvider = new RadioProvider({
                            isOn : {
                                get : function() {
                                    return false;
                                },
                                set : function(value) {}
                            }
                        }, dependencies);
                    }).not.toThrow();

                    expect(function() {
                        var radioProvider = new RadioProvider({
                            addFavoriteStation : function(opArgs) {
                                return true;
                            }
                        }, dependencies);
                    }).not.toThrow();
                });

        it("RadioProvider is instantiable", function() {
            var radioProvider = new RadioProvider({}, dependencies);
            expect(radioProvider).toBeDefined();
            expect(radioProvider).not.toBeNull();
            expect(typeof radioProvider === "object").toBeTruthy();
            expect(radioProvider instanceof RadioProvider).toBeTruthy();
        });

        it("RadioProvider has all members", function() {
            var radioProvider = new RadioProvider({}, dependencies);
            expect(radioProvider.isOn).toBeDefined();
            expect(radioProvider.isOn instanceof ProviderAttributeNotifyReadWrite).toBeTruthy();
            expect(radioProvider.enumAttribute).toBeDefined();
            expect(radioProvider.enumAttribute instanceof ProviderAttributeNotifyReadWrite)
                    .toBeTruthy();
            expect(radioProvider.enumArrayAttribute).toBeDefined();
            expect(radioProvider.enumArrayAttribute instanceof ProviderAttributeNotifyReadWrite)
                    .toBeTruthy();
            expect(radioProvider.addFavoriteStation).toBeDefined();
            expect(radioProvider.addFavoriteStation instanceof ProviderOperation).toBeTruthy();
            expect(radioProvider.operationWithEnumsAsInputAndOutput).toBeDefined();
            expect(radioProvider.operationWithEnumsAsInputAndOutput instanceof ProviderOperation)
                    .toBeTruthy();
            expect(radioProvider.weakSignal).toBeDefined();
            expect(radioProvider.weakSignal instanceof ProviderEvent).toBeTruthy();
            expect(radioProvider.interfaceName).toBeDefined();
            expect(typeof radioProvider.interfaceName).toEqual("string");
        });

        it("RadioProvider recognizes a correct implementation", function() {
            var radioProvider = new RadioProvider(implementation, dependencies);
            expect(radioProvider.checkImplementation).toBeDefined();
            expect(radioProvider.checkImplementation()).toBeTruthy();
            expect(radioProvider.checkImplementation().length).toEqual(0);
        });

        it("RadioProvider recognizes an incorrect implementation", function() {
            delete implementation.addFavoriteStation;
            var radioProvider = new RadioProvider(implementation, dependencies);
            expect(radioProvider.checkImplementation).toBeDefined();
            expect(radioProvider.checkImplementation()).toBeTruthy();
            expect(radioProvider.checkImplementation().length).toEqual(1);
        });

        it("RadioProvider recognizes a missing getter", function() {
            delete implementation.isOn.get;
            var radioProvider = new RadioProvider(implementation, dependencies);
            expect(radioProvider.checkImplementation).toBeDefined();
            expect(radioProvider.checkImplementation()).toBeTruthy();
            expect(radioProvider.checkImplementation().length).toEqual(1);
        });

        it("RadioProvider recognizes a missing setter", function() {
            delete implementation.isOn.set;
            var radioProvider = new RadioProvider(implementation, dependencies);
            expect(radioProvider.checkImplementation).toBeDefined();
            expect(radioProvider.checkImplementation()).toBeTruthy();
            expect(radioProvider.checkImplementation().length).toEqual(1);
        });

        it("RadioProvider recognizes multiple missing implementation functions", function() {
            delete implementation.isOn.get;
            delete implementation.addFavoriteStation;
            var radioProvider = new RadioProvider(implementation, dependencies);
            expect(radioProvider.checkImplementation).toBeDefined();
            expect(radioProvider.checkImplementation()).toBeTruthy();
            expect(radioProvider.checkImplementation().length).toEqual(2);
        });

    });

}); // require

File: java/core/libjoynr/src/test/java/io/joynr/dispatching/TestProvider.java
/*
 * #%L
 * %%
 * Copyright (C) 2011 - 2017 BMW Car IT GmbH
 * %%
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 * #L%
 */
package io.joynr.dispatching;

import static org.junit.Assert.assertEquals;
import static org.junit.Assert.assertTrue;

import java.util.Collection;
import java.util.HashMap;
import java.util.Map;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import io.joynr.provider.Promise;
import joynr.Request;
import joynr.tests.DefaulttestProvider;

public class TestProvider extends DefaulttestProvider {

    // maps from request payload to response
    private Map<Object, Object> sentPayloads = new HashMap<Object, Object>();
    private static final Logger logger = LoggerFactory.getLogger(TestProvider.class);
    private WaitTillCondition waitTillCondition;

    public TestProvider(int numberOfMessagesExpected) {
        waitTillCondition = new WaitTillCondition(numberOfMessagesExpected) {

            @Override
            protected Collection<Object> getReceivedPayloads() {
                return sentPayloads.keySet();
            }
        };
    }

    @Override
    public Promise<MethodWithStringsDeferred> methodWithStrings(String payload) {

        MethodWithStringsDeferred deferred = new MethodWithStringsDeferred();
        logger.info("Responding to payload: {}", payload.toString());
        String response = "response to " + payload.toString();
        sentPayloads.put(payload, response);

        waitTillCondition.releaseSemaphorePermit();
        deferred.resolve(response);
        return new Promise<>(deferred);
    }

    public Collection<Object> getSentPayloads() {
        return sentPayloads.values();
    }

    public String getSentPayloadFor(String payload) {
        return sentPayloads.get(payload).toString();
    }

    public Object getSentPayloadFor(Request request) {
        String payload = request.getParams()[0].toString();
        Object sentPayload = sentPayloads.get(payload);
        return sentPayload;
    }

    public void assertSentPayloadsContains(Object... payloads) {
        assertEquals(payloads.length, getSentPayloads().size());
        for (Object payload : payloads) {
            assertTrue(getSentPayloads().contains(payload));
        }
    }

    public void waitForMessage(int timeOutMs) {
        waitTillCondition.waitForMessage(timeOutMs);
    }

    public void assertAllPayloadsReceived(int timeOutMs) {
        waitTillCondition.assertAllPayloadsReceived(timeOutMs);
    }

    public void assertReceivedPayloadsContainsNot(String payloads) {
        waitTillCondition.assertReceivedPayloadsContainsNot(payloads);
    }

    public void assertReceivedPayloadsContains(String payloads) {
        waitTillCondition.assertReceivedPayloadsContains(payloads);
    }
}
